---
title: "DADA2 GWAS Replication Preparation"
author: "David A Hughes"
date: "28 June 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## A step by step to prepare DADA2 microbiome count data for FGFP GWAS replication


### load your data

```{r}
## source the FUNCTIONS file.
source("DADA2_DataSet_4_FGFP_FUNCTIONS.R")

####################################
## directory holding your data
## which should be 5 .tsv files
## that include the names
## Phylum, Class, Order, Family, Genus
datadir = c("yourcompletedirectorypath")

## make sure the provided directory path ends with a "/"
char = strsplit(datadir, split = "")[[1]]
if(char[length(char)] != "/"){
  datadir = paste0( paste0(char, collapse = ""), "/" )
  }

## Read file names in datadir
datafiles = list.files(datadir)

## Taxanomic Levels
taxalevels = c("Phylum","Class","Order","Family","Genus")

## Order data files by taxanomic level
o = sapply(taxalevels, function(x){ grep(x, datafiles) }) 
datafiles =  datafiles[o]

## Read in the data
mydata = lapply(datafiles, function(f){
    d = read.table(paste0(datadir, f ), h = TRUE, sep = "\t")
  })
names(mydata) = taxalevels

## set column 1, which should be taxa names as row names
## then eleminate column 1 from matrix
for(i in 1:length(mydata)){
  rownames(mydata[[i]]) = paste0(mydata[[i]][,1] )
  mydata[[i]] = mydata[[i]][,-1]
  }

## Edit unclassified taxa names
Tlevel = c("P","C","O","F","G")
for(i in 1:length(mydata)){
  n = paste0( rownames(mydata[[i]]) )
  n = gsub("unclassified", paste0( Tlevel[i], "_unclassified") , n)
  ##
  rownames(mydata[[i]]) = n
  }

## rotate mydata matrix to place individuals into rows
for(i in 1:length(mydata)){
  mydata[[i]] = t(mydata[[i]])
  }

## Matrix sizes ?
lapply(mydata, dim)
```


# FOR FGFP ONLY

```{r}
dir = "yourcompletedirectorypath"
## lab linkers
linker = read.table( paste0(dir, "LABID_VDP.list.txt"), h = TRUE)
## genetic ids
genids = read.table( paste0(dir, "fgfp_samples_ids.txt"), h = TRUE)

## Edit lab gen id
linker[,1] = gsub("\\.",".0",linker[,1])
## Add Genetic ID to the linker table
m = match(linker[,1], genids[, 2])
linker$genid = genids[m,1]
linker$inclusion = genids[m,3]
#############################
## ID SWAP on count tables
#############################
## Step 1: add GenID
for(i in 1:length(mydata)){
    m = match(rownames(mydata[[i]]), linker[,2] )
    rownames(mydata[[i]]) = linker[m, "genid"]
  }
```


## Load Covariate Data

```{r}
##### THIS WILL BE NEED TO BE MODIFIED FOR YOUR STUDY SPECIFICALY !
##### Covar data should include:
# 1) IDs, SEX, AGE, PC1-PC10
# 2) Batch effects: particularly those with observational associations to taxa abundances
#    ours were extract type, extract year, aliquot year, aliquot by, PCR plate. 

## I am loading data from a previous bgen data.sample file
covar_samplefile = "yourcompletedirectorypath/FGFP_dada2_v3_PRIMARY_data.sample"
covar_data = read.table( covar_samplefile, header = TRUE, sep = " ")
covar_data = covar_data[, 1:21]

## extract variable type (first row of data.sample file)
sampledata_variabletype = covar_data[1, ]

## NOTE: D = discrete covar; C = continuous covar; P = Continuous pheno; B = Binary pheno
## format info: http://www.stats.ox.ac.uk/~marchini/software/gwas/file_format.html

## remove variable type
covar_data = as.data.frame( covar_data[-1,] )

### STRUCTURE check
# str(covar_data)
cols2numeric = c(6:16)
## modify those columns to numeric that are necessary
for(i in 1:length(cols2numeric)){
  covar_data[, cols2numeric[i] ] = as.numeric( as.character(covar_data[,cols2numeric[i]  ]) )
  }

rownames(covar_data) = covar_data[,1]

# str(covar_data)

```


## Order microbiome abundances to line up with covar_data

*I am assuming that the top 10PC will be in the covar_data, thus the covar_data sample IDs will dictate the appropriate sample size. I.E. I assume any samples with out genotype data is not present in the covar_data file, and thus will be excluded by the ID matching step below.*

```{r}
for(i in 1:length(mydata)){
  m = match( covar_data[, "ID_1"], rownames(mydata[[i]]))
  mydata[[i]] = mydata[[i]][m,]
}

## matrix sizes ?
lapply(mydata, dim)
```



### Identify any potential outliers and turn to NAs

#### Bray Curtis Distances and MDS estimation

```{r}
## set genera data
gdata = mydata[["Genus"]] 
## estimate bray curtis distances
disMat <- vegdist(gdata, method="bray")
## a quick MDS
iMDS <- isoMDS(disMat)
```

#### A Shepard plot with ordination distance plotted against dissimilarity. 

```{r, fig.width = 5, fig.height=4 }
stressplot(iMDS, disMat)
```

#### Plot quick MDS

```{r, fig.width = 6, fig.height = 6}
### Mean and SD of MDS
### outliers are 5*SD from the mean
msd = c( mean(iMDS[[1]][,1]),  sd(iMDS[[1]][,1]) )
cuttoff_1 = c ( msd[1]+(5*msd[2] ) , msd[1]-(5*msd[2]) )
msd = c( mean(iMDS[[1]][,2]),  sd(iMDS[[1]][,2]) )
cuttoff_2 = c ( msd[1]+(5*msd[2] ) , msd[1]-(5*msd[2]) )
##
cuttoffs = cbind(cuttoff_1, cuttoff_2)
    
## PLOT
#postscript("SampleQC_isoMDS.eps", width = 8, height = 8)
###
plot( iMDS[[1]][,1] , iMDS[[1]][,2], pch = 21, bg = "dodgerblue", cex = 1.5,
      xlab = "Dim 1", ylab = "Dim 2", main = "dada2 all genera isoMDS")
abline(v = cuttoffs[,1] , col = "red")
abline(h = cuttoffs[,2] , col = "red")
### annotate outliers
outliers = which(iMDS[[1]][,1] > cuttoffs[1,1] | iMDS[[1]][,1] < cuttoffs[2,1] | iMDS[[1]][,2] > cuttoffs[1,2] | iMDS[[1]][,2] < cuttoffs[2,2] )
##
text(iMDS[[1]][outliers,1] , iMDS[[1]][outliers,2], 
     rownames(iMDS[[1]])[outliers], pos = 1, cex = 0.5)
###
#dev.off()

```

#### Turn outlier data into NAs

```{r}
if(length(outliers) > 0){
  for(i in 1:length(mydata)){
    mydata[[i]][ outliers,] = NA
    }
}
```



#### Estimate Summary Statistics on Taxa Data

```{r}
sumstats = lapply(mydata, function(mat){
    ss_out = t( sapply(1:ncol(mat), function(i){
        #print(i)
        out = TaxaSumStatFun( mat[,i] )
        return(out)
      }) )
    taxa = colnames(mat)
    rownames(ss_out) = taxa
    ss_out = cbind(taxa, ss_out)
    return(ss_out)
  })


## Add taxa level to table
for(i in 1:length(taxalevels)){
  taxalevel = rep(taxalevels[i], nrow(sumstats[[i]]) )
  sumstats[[i]] = cbind( taxalevel, sumstats[[i]] )
  }

## Pace all data into a single table
Allsumstats = c()
for(i in 1:length(sumstats)){
  Allsumstats = rbind(Allsumstats, sumstats[[i]] )
}
Allsumstats = as.data.frame(Allsumstats)
for(i in 3:ncol(Allsumstats)){
  Allsumstats[,i] = as.numeric(Allsumstats[,i])
  }

## Write Summary Statistics to file
write.table(Allsumstats, file = "Allsumstats.txt", row.names = TRUE, col.names = TRUE,
            sep = "\t", quote = FALSE)
```



## Estimate Alpha Diversity Stats

```{r, fig.width = 10, fig.height = 3.5}
## isolate Genus level data
gendata =  mydata[["Genus"]]
sampleids = rownames(gendata)

## ALPHA diversity
Div_NumberGenera = apply( gendata[-outliers, ], 1, function(x){ sum(x > 0) })
Div_Shannon = diversity( gendata, index = "shannon", MARGIN = 1, base = exp(1))
## 
chao = t( estimateR( gendata ) )
Div_Chao1 = chao[,2]

## id matching and sample order consistencey
m = match(sampleids, names(Div_NumberGenera))
Div_NumberGenera = Div_NumberGenera[m]; names(Div_NumberGenera) = sampleids
Div_Shannon = Div_Shannon[m]; names(Div_Shannon) = sampleids
Div_Chao1 = Div_Chao1[m]; names(Div_Chao1) = sampleids


### Plot Alpha Diversity
par(mfrow = c(1,3))
hist(Div_NumberGenera, col = "skyblue")
hist(Div_Shannon, col = "dodgerblue")
hist(Div_Chao1, col = "royalblue")
```

## Estimate Beta Diversity

```{r}
gendata =  mydata[["Genus"]]
gendata = as.data.frame( gendata )
sampleids = rownames(gendata)
###
MDS = metaMDS( as.data.frame( gendata[-outliers, ] ), distance = "bray", k = 2, try = 20, trymax = 50,  trace = FALSE)
betadiv_mds = MDS$points

## id matching and sample order consistencey
m = match(sampleids, rownames(betadiv_mds))
betadiv_mds = betadiv_mds[m,]
rownames(betadiv_mds) = sampleids
```

### Combine Alpha and Beta Diversity

```{r}
AB_diversity = cbind(Div_NumberGenera, Div_Shannon, Div_Chao1, betadiv_mds)
```

## Plot MDS with individuals and genus data

```{r, fig.width = 7, fig.height = 6}
library(RColorBrewer)
pcol = brewer.pal(9, "Spectral")
## plot limits
xminmax = c(  min( c(MDS$points[,1], na.omit(MDS$species[,1]) ) ) , 
              max( c(MDS$points[,1], na.omit(MDS$species[,1]) ) ) )

yminmax = c(  min( c(MDS$points[,2], na.omit(MDS$species[,2]) ) ) , 
              max( c(MDS$points[,2], na.omit(MDS$species[,2]) ) ) )             

### plot
plot(MDS$points[,1], MDS$points[,2], pch = 21, bg = pcol[9],
     xlim = xminmax, ylim = yminmax,cex = 1.5,
     xlab = "NMDS1", ylab = "NMDS2",
     sub = paste0("stress = ", signif(MDS$stress, d = 3)))
par(new = TRUE)
plot(MDS$species[,1], MDS$species[,2], pch = 22, bg = pcol[7],
     xlim = xminmax, ylim = yminmax, cex = 1,
     ann = FALSE, xaxt = "n", yaxt = "n")
###
msd = c(mean( na.omit(MDS$species[,1])), sd(na.omit(MDS$species[,1])))
k1 = which(MDS$species[,1] > msd[1]+(1.5*msd[2]) | MDS$species[,1] < msd[1]-(1.5*msd[2]))
k2 = which(MDS$species[,2] > msd[1]+(1.5*msd[2]) | MDS$species[,2] < msd[1]-(1.5*msd[2]))
w = unique(c(k1,k2))
par(new = TRUE)
text(  MDS$species[w,1], MDS$species[w,2], labels = names(MDS$species[w,1]) , xlim = xminmax, ylim = yminmax, cex = 0.45, pos = 3)

```

#### Load taxa that need be GWAS analyzed

```{r}
GWAStqxafile = "analyzable_taxa.txt"
GWAShurdletqxafile = "hurdletaxa.txt"

GWAStaxa = read.table(GWAStqxafile, h = FALSE, as.is = TRUE)[,1]
GWASHURDLEtaxa = read.table(GWAShurdletqxafile, h = FALSE, as.is = TRUE)[,1]

```

#### Complete Taxa Table

```{r}
alltaxa = c()
for(i in 1:length(mydata)){
  alltaxa = cbind( alltaxa, mydata[[i]] )
}
## add alpha and beta diversity
alltaxa = cbind(alltaxa, AB_diversity)

## edit name
colnames(alltaxa) = gsub("G_Escherichia/Shigella", "G_Escherichia_Shigella" ,  colnames(alltaxa) )


## extract only the taxa to analyze
m = match(GWAStaxa, colnames(alltaxa))
gwastaxadata = alltaxa[, m]

dim(gwastaxadata) ## 95 taxa 
```

## Transformations

```{r}
## We will perform a LOG and a RNT Transformation
## conditioned on the necessity of the sample going through
## a Hurdle step or not. If it is a Hurdle taxa, zero values
## will be truncated and all non-zero values will then me
## logged or rntransformed
gwastaxadata_transformed = c()
  for(i in 1:ncol(gwastaxadata)){
    ## taxa name
    n = colnames(gwastaxadata)[i]
    ## abundance data
    x = gwastaxadata[,i]
    original = x
    ### is it a hurdle taxa ??
    h = sum(n %in% GWASHURDLEtaxa)
    if(h == 1){
      ## set hb data
      hb = x
      hb[hb > 0] = 1
      ## truncated abundance
      x[x == 0] = NA
      ## Rank Normal transform
      rnt = rntransform(x)
      ## log2 transform
      ## must turn 0 into NAs for log2 transform
      log = log2(x)
      ## data out
      out = cbind(original, rnt, log, hb)
      colnames(out) = paste0( n, c("" ,"_RNT", "_LOG" , "_HB" ) )
    } else {
        ## Rank Normal transform
        rnt = rntransform(x)
        ## log2 transform
        ## must turn 0 into NAs for log2 transform
        x[x == 0] = NA
        log = log2(x)
        out = cbind(original, rnt, log)
        colnames(out) = paste0( n, c("" ,"_RNT", "_LOG" ) )
    }
    gwastaxadata_transformed = cbind(gwastaxadata_transformed, out)

    }

```


## Fit the RNT and LOG data to your covariates in a linear model and extract the residuals. 

```{r}
## identify those RNT and LOG transformed taxa
RNTLOG = sort( 
  c( grep("RNT", colnames(gwastaxadata_transformed)),
     grep("LOG", colnames(gwastaxadata_transformed))) 
                )
## make a vector of your sample IDs
mysamples = rownames(gwastaxadata_transformed)

## fit the RNT and LOG taxa to a lm()
ResTaxaData = sapply(RNTLOG, function(x){
  ## abundance data
  transformed_counts = gwastaxadata_transformed[,x]
  ## fit to lm()
  ## MAKE THE MODEL YOUR OWN - just be sure to include age, sex and top10PCs
  fit = lm( transformed_counts ~ 1  + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + exttype + extractyear + aliquotyear + aliquotby + plate + snpsex + age, data = covar_data)
  ## extract residuals
  res = residuals(fit)[mysamples]
  ## return residuals
  return( res )
  }) 
colnames(ResTaxaData) = paste0( colnames(gwastaxadata_transformed)[RNTLOG], "res" )

## write everytingto file
all = cbind(covar_data, gwastaxadata_transformed, ResTaxaData)
write.table(all, file = "alldata4analyzabletaxa.txt", row.names = TRUE, col.names = TRUE, sep = "\t", quote = FALSE )

```

## Merge Residual Data with covar_data and HB data

NOTE: covar_data must have for its first three columns
ID_1 ID_2 missing

```{r}
hbt = grep("_HB", colnames(gwastaxadata_transformed) )
gwas_data = cbind( covar_data, ResTaxaData, gwastaxadata_transformed[, hbt] )
### add variable type
# NOTE: sampledata_variabletype comes from row one of the data.sample file I loaded as a covariate data file.
type = c(sampledata_variabletype, rep( "P" , ncol(ResTaxaData) ), rep("B", length(hbt) ) )
names(type) = colnames(gwas_data)
###
gwas_data = rbind(type, gwas_data)
```

## Write new data.sample file

```{r}
write.table(gwas_data, "dada2_transformed_residualized_gwasready_data.sample", row.names = FALSE, col.names = TRUE, sep = " ", quote = FALSE)
```







